README.md

[![Arduino CI](https://github.com/RobTillaart/DEVNULL/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/DEVNULL/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/DEVNULL/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/DEVNULL/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/DEVNULL/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/DEVNULL/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/DEVNULL.svg?maxAge=3600)](https://github.com/RobTillaart/DEVNULL/releases)


# DEVNULL

Arduino library for a /dev/null stream.


## Description

The library implements a stream class that mimics the **/dev/null**
device of a Linux system. You can write everything to it. 
You can never read data from it as it always returns EOF (end of file);

The 0.1.0 version is a minimal implementation that can be optimized.
it implements the **write(const uint8_t data)** call and e.g. a float is still
converted to individual characters that are send one after another.
Strings and text send every byte.

The advantage is that printing takes time and e.g. one can use it to measure 
the **print** class performance.

Performance can be increased by implementing all methods of the print interface
with only a return 0; (or at least **print(Type)** as the **println(T)** would only 
call once extra for the "\n".


## Versions

#### 0.1.2

- sets the timeout for reading to 0. No need to wait longer with DEVNULL.
  this improves the **find(...)** calls substantially.
- added **size_t write( const uint8_t \*buffer, size_t size)** for faster string processing.

#### 0.1.4

- add **lastByte()** to check last byte written.
- split of .cpp

## Interface

- **DEVNULL()** constructor, sets the timeout to zero.
- **int available()** always return zero.
- **int peek()** always returns EOF.
- **int read()** always return EOF.
- **void flush()** does nothing but keeps some compilers happy.
- **size_t write(const uint8_t data)** implements print interface. returns 1.
- **size_t write( const uint8_t \*buffer, size_t size)** implements print interface.
Returns size.
- **int lastByte()** returns last byte written (debug and test purpose).


## Operation

Use with care.

See examples.


## Future

#### Could

- add byte counter (uint32_t)
- investigate if DEVNULL can be used to harvest entropy?
  - sum xor of all data + timestamp?
  - enable / disable flag (complex)
  - => /dev/entropy class?


#### Wont

- add delay to mimic pause / tune behaviour for simulating devices
  - microseconds
  - delay per byte, esp long arrays might need other performance
  - out of scope for /dev/null => separate class?
- implement Print class to increase performance?
  - derived class?



CHANGELOG.md
# Change Log DEVFULL

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.1.5] - 2022-10-31
- add changelog.md
- add rp2040 to build-CI


## [0.1.4] - 2022-09-21
- split up .h in .cpp and .h
- add **lastByte()** to check last byte written.

## [0.1.3] - 2021-12-15
- update library.json, license, minor edits

## [0.1.2] - 2021-11-24
- update build-CI
- add badges
- sets the timeout for reading to 0. No need to wait longer with DEVNULL.
  this improves the **find(...)** calls substantially.
- added **size_t write( const uint8_t \*buffer, size_t size)** for faster string processing.

## [0.1.1] - 2020-12-18
- add Arduino-CI.


## [0.1.0] - 2020-06-23
- initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-12-18
// PURPOSE: unit tests for the DEVNULL library
//          https://github.com/RobTillaart/
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)
// assertNotNull(actual)

#include <ArduinoUnitTests.h>


#include "DEVNULL.h"


unittest_setup()
{
  fprintf(stderr, "DEVNULL_LIB_VERSION: %s\n", (char *) DEVNULL_LIB_VERSION);
}

unittest_teardown()
{
}


unittest(test_all)
{
  DEVNULL dn;

  assertEqual(0,   dn.available());
  assertEqual(EOF, dn.peek());
  assertEqual(EOF, dn.read());

  assertEqual(1,   dn.write('a'));
  assertEqual('a', dn.lastByte());

  assertEqual(11,  dn.print("hello world"));
  assertEqual('d', dn.lastByte());
  assertEqual(13,  dn.println("hello world"));
  assertEqual(13,  dn.lastByte());  // 13 == newline
}

unittest_main()

// --------


DEVNULL.h
#pragma once
//
//    FILE: DEVNULL.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.4
// PURPOSE: Arduino library for a /dev/null stream - useful for testing
//     URL: https://github.com/RobTillaart/DEVNULL
//
// HISTORY:
//  0.1.0   2020-06-23  initial version.
//  0.1.1   2020-12-18  add Arduino-CI.
//  0.1.2   2021-11-24  update build-CI, badges, etc.
//                      added write(data, length) + _timeOut.
//  0.1.3   2021-12-15  update library.json, license, minor edits



#include "Arduino.h"

#define DEVNULL_LIB_VERSION     (F("0.1.4"))


class DEVNULL : public Stream
{
public:
  DEVNULL();

  int    available();
  int    peek();
  int    read();
  void   flush();  //  placeholder to keep CI happy

  size_t write(const uint8_t data);
  size_t write( const uint8_t *buffer, size_t size);

  int    lastByte();

private:
  uint8_t  _bottomLessPit;
};


// -- END OF FILE --

